### ISSUE TITLE
[bug] Switching on/off local ca causes certificates to take numeric suffix

### ISSUE THREAD

**Issue by robertsLando** [JonasAlfredsson/docker-nginx-certbot#142]
I'm using this container paired with a web ui that allows users to configure their custom domain to use with certbot. By default i set `USE_LOCAL_CA` to 1 so certificates are auto-generated and then when the user set his own domain he can disable it.

If I have `USE_LOCAL_CA` set with domain `mydomain.com` and then I restart nginx with `USE_LOCAL_CA` disabled it will create certificates to `live/mydomain.com-000X` (where X is an incremental integer that increments on every renew). Removing all the content inside live folder and restarting fixes the issue. Wondering if it could make sense to add a check in run certbot script that automatically does this when it see there is a conflict, when this happen the configuration is broken as it points to not/existing certificates

**Comment by robertsLando** [comment]
> Then I would like to ask why you need to create a certificate ?before? the user has entered their domain.

When the machine starts it should work even with no configuration provided, then the user will enter its custom domain, this is why I use local ca by default enabled and let user disable it once it set his own domain.

> Furthermore, there is this sentence in the [Local CA](https://github.com/JonasAlfredsson/docker-nginx-certbot/blob/master/docs/advanced_usage.md?rgh-link-date=2022-09-21T13%3A48%3A58Z#local-ca) section just because there might be this kind of weird behavior when switching back and forth.

I may have missed that sentence, it makes sense, so I should simply ensure that folder is empty when turning on/off local ca.

Thanks @JonasAlfredsson for your time

**Comment by JonasAlfredsson** [comment]
I am here to help in the extent I can, but I am still not sure I understand the setup process. 
You start a VM/Docker image that does not have a domain name assigned yet so you do not have a URL you can connect to (thus does not need a certificate?), the user then enters a domain which then restarts the container with a proper nginx config with correct domain?

**Comment by robertsLando** [comment]
> You start a VM/Docker image that does not have a domain name assigned yet so you do not have a URL you can connect to (thus does not need a certificate?), the user then enters a domain which then restarts the container with a proper nginx config with correct domain?

Exactly, I simply don't know the user that will use the application so I need to just have nginx-certbot start and create some dummy certs to serve the page over https, then once it is up and running it can be configured

**Comment by JonasAlfredsson** [comment]
But why do you need to serve a webpage on a server which does not have a domain yet? Can't you wait to start it until you know the domain?

**Comment by robertsLando** [comment]
> But why do you need to serve a webpage on a server which does not have a domain yet? Can't you wait to start it until you know the domain?

This is by design, the software is installed the as it is on all machines without knowing who will use it, then we don't want to ask users to manually modify env files but he can use UI to do so

**Comment by JonasAlfredsson** [comment]
Alright, so there is some kind of "default" nginx config that you use to redirect people to a setup location before restarting to the proper location?

Can you post it because I might have another idea.

**Comment by robertsLando** [comment]
Not just that, I also use [nginx envsubs](https://github.com/nginxinc/docker-nginx/blob/master/entrypoint/20-envsubst-on-templates.sh)

This is my nginx entry on compose:

```Dockerfile
nginx:
      image: jonasal/nginx-certbot:3.2.0
      logging: *default-logging
      container_name: "${PREFIX}vpnbox-nginx"
      volumes:
        - ./nginx/conf.d/:/etc/nginx/user_conf.d/ # nginx-certbot automatically symlink files inside this thir to conf.d folder on entrypoint
        - ./nginx/templates:/etc/nginx/templates # nginx uses envsubst in his entrypoint in order to generate conf from templates
        - ./nginx/ssl/:/etc/letsencrypt/ # nginx-certbot stores ssl certs here
        - ./nginx/etc/nginx.conf:/etc/nginx/nginx.conf
        - ./nginx/etc/proxy.conf:/etc/nginx/proxy.conf
        - ./nginx/5xx.html:/var/www/html/5xx.html
      ports:
       - "${HTTP_PORT}:80"
       - "${HTTPS_PORT}:443"
      environment:
        - DOMAIN=${DOMAIN:-localhost}
        - CERTBOT_EMAIL=${CERTBOT_EMAIL:-example@mail.it}
        - CERTBOT_AUTHENTICATOR=${CERTBOT_AUTHENTICATOR:-webroot}
        - USE_LOCAL_CA=${USE_LOCAL_CA:-1}
      restart: always
```


My `default.conf.template` file: 

```
server {
        listen 443 ssl http2;
        listen [::]:443 ssl http2;
        server_name ${DOMAIN};

        server_tokens off;

        # Load the certificate files.
        ssl_certificate         /etc/letsencrypt/live/${DOMAIN}/fullchain.pem;
        ssl_certificate_key     /etc/letsencrypt/live/${DOMAIN}/privkey.pem;
        ssl_trusted_certificate /etc/letsencrypt/live/${DOMAIN}/chain.pem;

        # Load the Diffie-Hellman parameter.
        ssl_dhparam /etc/letsencrypt/dhparams/dhparam.pem;

        ssl_buffer_size 8k;

        ssl_protocols TLSv1.2 TLSv1.1 TLSv1;
        ssl_prefer_server_ciphers on;

        ssl_ecdh_curve secp384r1;
        ssl_session_tickets off;

        resolver 8.8.8.8;

        error_page 500 502 503 504 /5xx.html;

        location = /5xx.html {
                root /var/www/html;
                internal;
        }

        location / {
                resolver 127.0.0.11; # use docker's DNS server
                set $api_server "api:3001"; # workaround for issue [emerg] host not found in upstream "api": https://stackoverflow.com/a/52319161
                proxy_pass https://$api_server;
                include /etc/nginx/proxy.conf;
                add_header X-Frame-Options "SAMEORIGIN" always;
                add_header X-XSS-Protection "1; mode=block" always;
                add_header X-Content-Type-Options "nosniff" always;
                add_header Referrer-Policy "no-referrer-when-downgrade" always;
                add_header Content-Security-Policy "default-src * data: 'unsafe-eval' 'unsafe-inline' blob:" always;
        }
}

```

**Comment by JonasAlfredsson** [comment]
Well, if you are using templating you could perhaps template it to use unencrypted port 80 during the first boot, and then overwrite it with the correct one on the reboot?

**Comment by robertsLando** [comment]
By default users should always (and only) use port 443. This cannot be changed as there are other devices that connect to that port.

**Comment by JonasAlfredsson** [comment]
I am still unsure of how this entire process works, so here is my understanding so far (please correct me if wrong :) )

1. The user navigates to a URL where some kind of GUI allows them to request a VM.
2. A VM is started and this can now be accessed (how? SSH? remote desktop via the website? HTTP traffic?)
3. On this VM a couple of Docker containers are started, one of which is the Nginx one which needs to be updated to the correct DOMAIN to function correctly. 
4. This is done by navigating to "localhost" (if I understand the config correctly), which then forwards all requests to api:3001 which is the control plane(?) where it is possible to change stuff. 
5. After the user clicks apply the Nginx container will restart with the correct DOMAIN set.

There are a few question marks in the statements above, so the following is based on assumptions.

If the VM can be accessed via remote desktop (and they are supposed to open a webrowser on it and navigate to localhost), I don't really see it necessary to have HTTPS. You could probably just point them directly to api:3001 if on localhost, but if HTTPS is really important then I guess the current setup is probably the best you can get with minimal work.

---

P.S. if HTTPS is important then you should drop support for anything lower than TLS1.2.

**Comment by robertsLando** [comment]
Kind of that, the thing is a little bit more complicated. Firstly this solution can run on dedicated physical server or in cloud (VPS).

> A VM is started and this can now be accessed

Once started it should be configurable using HTTPS. There is also ssh available but normal users should not use it, it's just for assitance purposes

> This is done by navigating to "localhost" (if I understand the config correctly), which then forwards all requests to api:3001 which is the control plane(?) where it is possible to change stuff.

Exactly

> After the user clicks apply the Nginx container will restart with the correct DOMAIN set.

Yes along with also any authenticator configurations and local ca flag

> If the VM can be accessed via remote desktop 

It cannot be accessed via remote desktop

> but if HTTPS is really important then I guess the current setup is probably the best you can get with minimal work.

Unfortunately yes, it is, we cannot use HTTP, also based on project requirements we have to use less ports as possible

> P.S. if HTTPS is important then you should drop support for anything lower than TLS1.2.

This would be awesome but this software is used to provide remote vpn access to devices, this devices also include some old ones that doesn't support TLS1.2, dropping it would make them not working at all :( Will be something we will fix in the near future but have to keep it for now

**Comment by JonasAlfredsson** [comment]
OK, so if this machine must use HTTPS and you have no way of knowing what domain it will get I guess the LOCAL_CA way is the best. 
Seems to be a quite convoluted setup, and I can't say I understand it fully, but I suppose if it works it works :)

Sorry to hear about the legacy support. Hope that get fixed as soon as possible :)
